### ISSUE TITLE
web: ISE: IrSimpleFunctionSymbolImpl is already bound: FUN LOCAL_FUNCTION_FOR_LAMBDA with Jetpack Compose Plugin during compileProductionLibraryKotlinJs

### ISSUE THREAD

**Issue by eymar** [JetBrains/compose-jb#1114]
https://youtrack.jetbrains.com/issue/KT-48425

> Unfortunately, I can't give you a small reproducer, only the causing full project:
> https://github.com/hfhbd/bootstrap-compose/tree/dependabot/gradle/org.jetbrains.compose-1.0.0-alpha4-build321 (included)
> 
> Project contains Jetpack Compose Web plugin: id("org.jetbrains.compose") version "1.0.0-alpha4-build321"
> Kotlin version: 1.5.21 (no update to 1.5.30 possible due to Compose)
> 
> Any possibility to include more outputs?
> 
> Only happens during compileProductionLibraryKotlinJs, normal compileKotlinJs finishes successful.

```
java.lang.IllegalStateException: IrSimpleFunctionSymbolImpl is already bound: FUN LOCAL_FUNCTION_FOR_LAMBDA name:<anonymous> visibility:local modality:FINAL <> () returnType:<Uninitialized>
	at org.jetbrains.kotlin.ir.symbols.impl.IrBindableSymbolBase.bind(IrPrivateSymbolBase.kt:64)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFunctionImpl.<init>(IrFunctionImpl.kt:99)
	at org.jetbrains.kotlin.ir.declarations.impl.IrFactoryImpl.createFunction(IrFactoryImpl.kt:125)
	at org.jetbrains.kotlin.ir.declarations.IrFactory$DefaultImpls.createFunction$default(IrFactory.kt:91)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$deserializeIrFunction$1$2.invoke(IrDeclarationDeserializer.kt:500)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$deserializeIrFunction$1$2.invoke(IrDeclarationDeserializer.kt:498)
	at org.jetbrains.kotlin.ir.util.SymbolTable.declareSimpleFunction(SymbolTable.kt:1803)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeIrFunction$ir_serialization_common(IrDeclarationDeserializer.kt:498)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeDeclaration(IrDeclarationDeserializer.kt:688)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:135)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlock(IrBodyDeserializer.kt:153)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:764)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeMemberAccessCommon(IrBodyDeserializer.kt:163)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeCall(IrBodyDeserializer.kt:272)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:767)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeMemberAccessCommon(IrBodyDeserializer.kt:163)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeCall(IrBodyDeserializer.kt:272)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:767)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBranch(IrBodyDeserializer.kt:103)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:131)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeWhen(IrBodyDeserializer.kt:607)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:796)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlock(IrBodyDeserializer.kt:153)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:764)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBranch(IrBodyDeserializer.kt:103)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:131)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeWhen(IrBodyDeserializer.kt:607)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:796)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlockBody(IrBodyDeserializer.kt:94)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:129)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeStatementBody(IrDeclarationDeserializer.kt:460)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$withDeserializedIrFunctionBase$1$1$1$1.invoke(IrDeclarationDeserializer.kt:487)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$withDeserializedIrFunctionBase$1$1$1$1.invoke(IrDeclarationDeserializer.kt:480)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.withBodyGuard(IrDeclarationDeserializer.kt:416)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.access$withBodyGuard(IrDeclarationDeserializer.kt:56)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeIrFunction$ir_serialization_common(IrDeclarationDeserializer.kt:892)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeFunctionExpression(IrBodyDeserializer.kt:333)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:801)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeMemberAccessCommon(IrBodyDeserializer.kt:163)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeCall(IrBodyDeserializer.kt:272)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:767)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeMemberAccessCommon(IrBodyDeserializer.kt:163)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeCall(IrBodyDeserializer.kt:272)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:767)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlock(IrBodyDeserializer.kt:153)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:764)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBranch(IrBodyDeserializer.kt:103)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:131)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeWhen(IrBodyDeserializer.kt:607)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:796)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlock(IrBodyDeserializer.kt:153)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:764)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlock(IrBodyDeserializer.kt:153)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:764)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBranch(IrBodyDeserializer.kt:103)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:131)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeWhen(IrBodyDeserializer.kt:607)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeOperation(IrBodyDeserializer.kt:796)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeExpression(IrBodyDeserializer.kt:813)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:137)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeBlockBody(IrBodyDeserializer.kt:94)
	at org.jetbrains.kotlin.backend.common.serialization.IrBodyDeserializer.deserializeStatement$ir_serialization_common(IrBodyDeserializer.kt:129)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeStatementBody(IrDeclarationDeserializer.kt:460)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$withDeserializedIrFunctionBase$1$1$1$1.invoke(IrDeclarationDeserializer.kt:487)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer$withDeserializedIrFunctionBase$1$1$1$1.invoke(IrDeclarationDeserializer.kt:480)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.withBodyGuard(IrDeclarationDeserializer.kt:416)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.access$withBodyGuard(IrDeclarationDeserializer.kt:56)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeIrFunction$ir_serialization_common(IrDeclarationDeserializer.kt:892)
	at org.jetbrains.kotlin.backend.common.serialization.IrDeclarationDeserializer.deserializeDeclaration(IrDeclarationDeserializer.kt:688)
	at org.jetbrains.kotlin.backend.common.serialization.IrFileDeserializer.deserializeDeclaration$ir_serialization_common(IrFileDeserializer.kt:36)
	at org.jetbrains.kotlin.backend.common.serialization.FileDeserializationState.deserializeAllFileReachableTopLevel(IrFileDeserializer.kt:127)
	at org.jetbrains.kotlin.backend.common.serialization.ModuleDeserializationState.deserializeReachableDeclarations(BasicIrModuleDeserializer.kt:163)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.deserializeAllReachableTopLevels(KotlinIrLinker.kt:105)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.findDeserializedDeclarationForSymbol(KotlinIrLinker.kt:124)
	at org.jetbrains.kotlin.backend.common.serialization.KotlinIrLinker.getDeclaration(KotlinIrLinker.kt:162)
	at org.jetbrains.kotlin.ir.util.ExternalDependenciesGeneratorKt.getDeclaration(ExternalDependenciesGenerator.kt:61)
	at org.jetbrains.kotlin.ir.util.ExternalDependenciesGenerator.generateUnboundSymbolsAsDependencies(ExternalDependenciesGenerator.kt:48)
	at org.jetbrains.kotlin.ir.backend.js.KlibKt.loadIr(klib.kt:288)
	at org.jetbrains.kotlin.ir.backend.js.CompilerKt.compile(compiler.kt:57)
	at org.jetbrains.kotlin.ir.backend.js.CompilerKt.compile$default(compiler.kt:36)
	at org.jetbrains.kotlin.cli.js.K2JsIrCompiler.doExecute(K2JsIrCompiler.kt:256)
	at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:182)
	at org.jetbrains.kotlin.cli.js.K2JSCompiler.doExecute(K2JSCompiler.java:75)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:90)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:76)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execAndOutputXml(CLICompiler.kt:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork.compileInProcessImpl(GradleKotlinCompilerWork.kt:347)
	at org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork.access$compileInProcessImpl(GradleKotlinCompilerWork.kt:75)
	at org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork$compileInProcess$future$1.call(GradleKotlinCompilerWork.kt:320)
	at org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork$compileInProcess$future$1.call(GradleKotlinCompilerWork.kt:319)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by ylemoigne** [comment]
Same issue here with build328 (and kotlin 1.5.30).
Can you explain how the workaround is working ? (or any way to locate the part that break the compiler)

**Comment by eymar** [comment]
cc: @ylemoigne 

It breaks when there is a call of `@Composable` lambda that takes another lambda as an argument.

In example above `control` type is `@Composable ElementScope<HTMLDivElement>.(List<Table.Pagination.Page<T>>, CurrentPage<T>, (Int) -> Unit) -> Unit`

```
val control = pagination.control
control(pages, currentPage) { // it's calling composable lambda
    currentIndex = it
}
```            

It will also break if code looks like this:
```
val control = pagination.control
control(pages, currentPage, someLambdaWrapper { ... }) 
```            
            
Therefore, lifting the lambda up into a variable and using that variable as an argument helps:
```
val control = pagination.control
val l : (Int) -> Unit = {
    currentIndex = it
}
control(pages, currentPage, l)
```

or
```
val control = pagination.control
val l = { ... }
control(pages, currentPage, someLambdaWrapper(l)) 
```            


Of course, it's a temporary workaround. The fix is on its way (probably not the soonest build though).

**Comment by ylemoigne** [comment]
Ok, thanks for quick reply 👍

**Comment by ylemoigne** [comment]
Compilation is ok for me with 0.0.0-SNAPSHOT (20210828.033914-25)

**Comment by hfhbd** [comment]
@ylemoigne What do you mean? It works with version < 321.

**Comment by ylemoigne** [comment]
Sorry, I mean it work **without workaround** with 0.0.0-SNAPSHOT (20210828.033914-25)

**Comment by hfhbd** [comment]
Yeah, I meant this too. It compiles without problems/workarounds in version 318, but in 321 it needs this workaround.

**Comment by ylemoigne** [comment]
And release before 321 works with kotlin 1.5.30 ?

**Comment by hfhbd** [comment]
@ylemoigne I don’t know, what you mean. Kotlin 1.5.30 is only supported in 328 or later.

**Comment by ylemoigne** [comment]
I'm sorry, I don't understand what are your intention.

Mine was just to inform people who arrives here that the issue is fixed (in the lastest unreleased build).

So now we (and people affected) know that :

| Release | Issue | Kotlin 1.5.30 Compat |
| --- | --- | --- |
| < 321 | No | No |
| 321 | Yes | No |
| 328 | Yes | Yes |
| >=20210828.033914  | No | Yes |

**Comment by eymar** [comment]
closing this issue. It was fixed in >= 1.0.0-alpha4-build331 versions

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-6300">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
